wimm.Secundatives/Extensions/MappingExtensions.cs

```diff
@@ -0,0 +1,104 @@
+﻿using System;
+using System.Threading.Tasks;
+
+namespace wimm.Secundatives
+{
+    public static class MappingExtensions
+    {
+
+
+        /// <summary>
+        /// Transforms a <see cref="Maybe{T}"/> into a <see cref="Maybe{U}"/> by applying a function in the case that <paramref name="value"/>
+        /// contains a value or <see cref="Maybe{U}.None"/> if <paramref name="value"/> is <see cref="Maybe{T}.None"/>
+        /// </summary>
+        /// <typeparam name="T"> The type to be mapped from</typeparam>
+        /// <typeparam name="U"> The type to be mapped to </typeparam>
+        /// <param name="value"> A <see cref="Maybe{T}"/> that determines if a transform occurs </param>
+        /// <param name="func"> The function that will transform <paramref name="value"/>'s T</param>
+        /// <returns> A <see cref="Maybe{U}"/> equal to <see cref="Maybe{U}.None"/> if <paramref name="value"/> is none otherwise a 
+        /// <see cref="Maybe{T}"/> containing the result of <paramref name="func"/> being applied to <paramref name="value"/>'s 
+        /// <typeparamref name="T"/>
+        /// </returns>
+        public static Maybe<U> Map<T, U>(this Maybe<T> value, Func<T, U> func)
+        {
+            return value.Exists
+                ? func(value.Value)
+                : Maybe<U>.None;
+        }
+
+        /// <summary>
+        /// Transforms a <see cref="Maybe{T}"/> into a <see cref="Maybe{U}"/> by applying a function in the case that <paramref name="value"/>
+        /// contains a value or <see cref="Maybe{U}.None"/> if <paramref name="value"/> is <see cref="Maybe{T}.None"/>
+        /// </summary>
+        /// <typeparam name="T"> The type to be mapped from</typeparam>
+        /// <typeparam name="U"> The type to be mapped to </typeparam>
+        /// <param name="value"> A <see cref="Maybe{T}"/> that determines if a transform occurs </param>
+        /// <param name="func"> A function that produces the resultant <see cref="Maybe{T}"/></param>
+        /// <returns> A <see cref="Maybe{U}"/> equal to <see cref="Maybe{U}.None"/> if <paramref name="value"/> is none
+        /// otherwise the result of <paramref name="func"/> being applied to <paramref name="value"/>'s <typeparamref name="T"/>
+        /// </returns>
+        public static Maybe<U> Map<T, U>(this Maybe<T> value, Func<T, Maybe<U>> func)
```

Comment:
does the existence of this method prevent an expression from resolving to the method above as `Maybe<Maybe{U}> Map<T, Maybe{U}>(this Maybe<T> value, Func<T, Maybe{U}>` ?

Reply:
Yeah, without this and the last one the flattening doesn't happen and you can have a Maybe<Maybe<Maybe<Maybe<U>>>>